lib/generators.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python3
+
+# SPDX-FileCopyrightText: 2009 Fermi Research Alliance, LLC
+# SPDX-License-Identifier: Apache-2.0
+
+#
+# Project:
+#   glideinWMS
+#
+# File Version:
+#
+# Description:
+#   Contains the Generator base class and built-in generators
+#
+
+import os
+import re
+import sys
+
+from abc import ABC, abstractmethod
+from typing import Generic, List, Mapping, Optional, TypeVar
+
+from glideinwms.lib.util import import_module
+
+sys.path.append("/etc/gwms-frontend/plugin.d")
+_loaded_generators = {}
+
+T = TypeVar("T")
+
+
+class Generator(ABC, Generic[T]):
+    def __str__(self):
+        return f"{self.__class__.__name__}()"
+
+    def __repr__(self):
+        return str(self)
+
+    @abstractmethod
+    def generate(self, arguments: Optional[Mapping] = None) -> T:
+        pass
+
+
+def load_generator(module: str) -> Generator:
+    """Load a generator from a module
+
+    Args:
+        module (str): module that exports a generator
+
+    Raises:
+        ImportError: when a `Generator` object cannot be imported from `module`
+
+    Returns:
+        Generator: generator object
+    """
+
+    module_name = re.sub(r"\.py[co]?$", "", os.path.basename(module))  # Extract module name from path
+
+    try:
+        if not module_name in _loaded_generators:
+            imported_module = import_module(module)
+            if module_name not in _loaded_generators:
```

Comment:
`_loaded_generators` is a module-wide variable. The `export_generator` function is called from a custom generator module, populating the `_loaded_generators` map. When `import_module(module)` is called, we expect `export_generator` to be called within the imported module, therefore updating `_loaded_generators`; otherwise, we raise an error.